pkg/plugins/manager/loader/finder/fs.go

```diff
@@ -0,0 +1,289 @@
+			continue
+		}
+
+		pluginJSONAbsPath, err := filepath.Abs(pluginJSONPath)
+		if err != nil {
+			f.log.Warn("Skipping plugin loading as absolute plugin.json path could not be calculated", "pluginID", plugin.ID, "err", err)
+			continue
+		}
+
+		if _, dupe := foundPlugins[filepath.Dir(pluginJSONAbsPath)]; dupe {
+			f.log.Warn("Skipping plugin loading as it's a duplicate", "pluginID", plugin.ID)
+			continue
+		}
+		foundPlugins[filepath.Dir(pluginJSONAbsPath)] = plugin
+	}
+
+	var res = make(map[string]*plugins.FoundBundle)
+	for pluginDir, data := range foundPlugins {
+		files, err := collectFilesWithin(pluginDir)
+		if err != nil {
+			return nil, err
+		}
+
+		res[pluginDir] = &plugins.FoundBundle{
+			Primary: plugins.FoundPlugin{
+				JSONData: data,
+				FS:       plugins.NewLocalFS(files, pluginDir),
+			},
+		}
+	}
+
+	var result []*plugins.FoundBundle
+	for dir := range foundPlugins {
+		ancestors := strings.Split(dir, string(filepath.Separator))
+		ancestors = ancestors[0 : len(ancestors)-1]
+
+		pluginPath := ""
+		if runtime.GOOS != "windows" && filepath.IsAbs(dir) {
+			pluginPath = "/"
+		}
+		add := true
+		for _, ancestor := range ancestors {
+			pluginPath = filepath.Join(pluginPath, ancestor)
+			if _, ok := foundPlugins[pluginPath]; ok {
+				if fp, exists := res[pluginPath]; exists {
+					fp.Children = append(fp.Children, &res[dir].Primary)
+					add = false
+					break
+				}
+			}
+		}
+		if add {
+			result = append(result, res[dir])
+		}
+	}
+
+	return result, nil
+}
+
+func (f *FS) getAbsPluginJSONPaths(path string) ([]string, error) {
+	var pluginJSONPaths []string
+
+	var err error
+	path, err = filepath.Abs(path)
+	if err != nil {
+		return []string{}, err
+	}
+
+	if err = walk(path, true, true,
+		func(currentPath string, fi os.FileInfo, err error) error {
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					f.log.Error("Couldn't scan directory since it doesn't exist", "pluginDir", path, "err", err)
+					return nil
+				}
+				if errors.Is(err, os.ErrPermission) {
+					f.log.Error("Couldn't scan directory due to lack of permissions", "pluginDir", path, "err", err)
+					return nil
+				}
+
+				return fmt.Errorf("filepath.Walk reported an error for %q: %w", currentPath, err)
+			}
+
+			if fi.Name() == "node_modules" {
+				return util.ErrWalkSkipDir
+			}
+
+			if fi.IsDir() {
+				return nil
+			}
+
+			if fi.Name() != "plugin.json" {
+				return nil
+			}
+
+			pluginJSONPaths = append(pluginJSONPaths, currentPath)
+			return nil
+		}); err != nil {
+		return []string{}, err
+	}
+
+	return pluginJSONPaths, nil
+}
+
+func (f *FS) readPluginJSON(pluginJSONPath string) (plugins.JSONData, error) {
+	f.log.Debug("Loading plugin", "path", pluginJSONPath)
+
+	if !strings.EqualFold(filepath.Ext(pluginJSONPath), ".json") {
+		return plugins.JSONData{}, ErrInvalidPluginJSONFilePath
+	}
+
+	// nolint:gosec
```

Comment:
Should be able to get rid of this using our earlier workaround

Reply:
Do you mean `util.CleanRelativePath(name)`?

Reply:
Partly, referring to https://github.com/grafana/grafana/blob/35629a1f077ebbed34c488bf82b9cec6f511bf91/pkg/plugins/plugins.go#L107